src/backend/cuda/join.cpp

```diff
@@ -11,98 +11,88 @@
-            odims[i] = fdims[i] + sdims[i];
-        } else {
-            odims[i] = fdims[i];
-        }
-    }
+    dim4 odims(fdims);
+    odims.dims[jdim] += sdims.dims[jdim];
 
     Array<T> out = createEmptyArray<T>(odims);
+    vector<kernel::CParamPlus<T>> ins;
+    ins.reserve(2);
+
+    if (first.isReady()) {
+        ins.emplace_back((CParam<T>)first, 0);
+    } else {
+        vector<Param<T>> outputs{{out.get(), fdims.dims, out.strides().dims}};
+        const vector<common::Node *> nodes{first.getNode().get()};
+        evalNodes<T>(outputs, nodes);
+    }
 
-    af::dim4 zero(0, 0, 0, 0);
-
-    kernel::join<T>(out, first, zero, dim);
-    kernel::join<T>(out, second, calcOffset(fdims, dim), dim);
+    if (second.isReady()) {
+        ins.emplace_back((CParam<T>)second,
+                         fdims.dims[jdim] * out.strides().dims[jdim]);
+    } else {
+        vector<Param<T>> outputs{
+            {out.get() + fdims.dims[jdim] * out.strides().dims[jdim],
+             sdims.dims, out.strides().dims}};
+        const vector<common::Node *> nodes{second.getNode().get()};
+        evalNodes<T>(outputs, nodes);
+    }
 
+    if (ins.size()) kernel::memcopyN<T>(out, ins);
     return out;
 }
 
 template<typename T>
-void join_wrapper(const int dim, Array<T> &out,
-                  const std::vector<Array<T>> &inputs) {
-    af::dim4 zero(0, 0, 0, 0);
-    af::dim4 d = zero;
-
-    kernel::join<T>(out, inputs[0], zero, dim);
-    for (size_t i = 1; i < inputs.size(); i++) {
-        d += inputs[i - 1].dims();
-        kernel::join<T>(out, inputs[i], calcOffset(d, dim), dim);
-    }
-}
-
-template<typename T>
-Array<T> join(const int dim, const std::vector<Array<T>> &inputs) {
+Array<T> join(const int jdim, const std::vector<Array<T>> &inputs) {
     // All dimensions except join dimension must be equal
     // Compute output dims
-    af::dim4 odims;
-    const dim_t n_arrays = inputs.size();
-    std::vector<af::dim4> idims(n_arrays);
-
-    dim_t dim_size = 0;
-    for (size_t i = 0; i < idims.size(); i++) {
-        idims[i] = inputs[i].dims();
-        dim_size += idims[i][dim];
+    dim4 odims(inputs[0].dims());
+    odims.dims[jdim] = 0;
+    for (auto &iArray : inputs) {
+        odims.dims[jdim] += iArray.dims().dims[jdim];
     }
 
-    for (int i = 0; i < 4; i++) {
-        if (i == dim) {
-            odims[i] = dim_size;
+    Array<T> out = createEmptyArray<T>(odims);
+    vector<kernel::CParamPlus<T>> ins;
+    ins.reserve(inputs.size());
+
+    dim_t outOffset = 0;
+    for (auto &iArray : inputs) {
+        if (iArray.isReady()) {
+            ins.emplace_back((CParam<T>)iArray, outOffset);
         } else {
-            odims[i] = idims[0][i];
+            vector<Param<T>> outputs{{out.get() + outOffset, iArray.dims().dims,
+                                      out.strides().dims}};
+            vector<common::Node *> nodes{iArray.getNode().get()};
+            evalNodes<T>(outputs, nodes);
```

Comment:
Again. I really like this. I think you can use evalMultiple here right?

I can see cases where the kernels may get too large but I am working on a way around that.

Reply:
I toughed that for evalMultiple(), all the arrays needed to be the same size, which is not guaranteed here.